### ISSUE TITLE
cannot resume stopped episode

### ISSUE THREAD

**Issue by sorin-costea** [popcorntime/popcorntime#1988]
I'm on Windows 10 and have the latest version dowloaded from Jenkins. 

I play an episode with about a dozen of seeds (and more peers), I press ESC and get out of it. I play it again, it would play what was already downloaded but stays stuck at 0 peers and 0 seeds. I've read some closed issues but:

- I think after waiting 5 minutes is reasonable to assume nothing will ever happen
- I deleted all the items in "seedbox", now stuck at "connecting"
- unchecking the "seedbox" feature in the settings only hid the button, no change in behaviour
- I deleted the cached files, no change
- there's a setting "activate udate seeding" which doesn't change anything either
- streaming or locally watching doesn't make a difference

![Unbenannt](https://user-images.githubusercontent.com/3449432/105265508-32775900-5b91-11eb-9681-16b28956f1ba.png)

So basically my only chance to watch an episode is to not ever interrupt it. Other episodes will work, unless I interrupt them - then they go dead, never ever managing to (re)connect. Is there some torrent intrinsic feature which kicks me out in some way?

**Comment by sorin-costea** [comment]
Correction!
It will reconnect and work, AFTER I RESTART PCT.
But it will still show weird values in the view popup, see screenshot.
![Unbenannt1](https://user-images.githubusercontent.com/3449432/105268157-f8f31d80-5b91-11eb-9ba9-385d63e1cf99.png)

**Comment by kiriles90** [comment]
Its an old issue related to how the Seedbox feature stops (pauses) torrents. More specifically 
https://github.com/popcorn-official/popcorn-desktop/blob/a5025d176141eb228e0b57009e7377962e4d48c7/src/app/lib/streamer.js#L130  

Which removes the peers then takes a good amount of time for them to be resolved again when resume (for some bad health torrents they dont resolve at all). Not sure what can be done about it since WebTorrent doesn't really pause connections with just `pause()` alone, but since curiously none has opened an issue about it (though it has been mentioned a number of times in other issues) marking this as 'Bug' in the hope that a workaround will be found.

For now if you want try disabling the Seedbox feature in the settings, Like  
https://github.com/popcorn-official/popcorn-desktop/blob/a5025d176141eb228e0b57009e7377962e4d48c7/src/app/lib/streamer.js#L126
implies the action taken to stop a torrent is different when the Seedbox feature is disabled. No `removePeer()` involved. Instead it defaults to the old, pre-seedbox, 1 active torrent Popcorn Time streamer behavior.

**Comment by sorin-costea** [comment]
Thank you!
Although there's still something weird, not obviously seedbox-related:
- that after the restart the Download/Upload numbers show 0 although it works fine
- that even if I delete the entry in the seedbox and the files from the cache directory it still won't work until a restart

**Comment by tkSimon** [comment]
It seems the main problem is that pct is fiddling with webtorrent internals (removing peers, canceling requests). Peers are added by discovery only after initializing a new WebTorrent instance, but not on resume. (Apart from peer discovery, which assumes there are peers connected to discover other peers from)
From what I understand, if there are more than maxConn peers available at the time of calling pause(), those might be used after resume, as they're not in the _peers array yet, but in the queue, which pct's pause() won't touch.

I think there are 3 options:
1. Don't mess with WebTorrent internals. This means that when stopping the player or loading screen the torrent would continue to download for some time - if seedbox is enabled - but would eventually stop.
2. Fully destroy the WebTorrent instance. I haven't explored this option, but i think this would be acceptable. When the user starts the player again it'll just create a new instance, with working discovery, and whatever was previously downloaded is still in the cache anyway, right?
3. When removing peers, store them somewhere, then when resuming, call addPeer on the torrent with the stored peers. I made a POC of that here: https://github.com/tkSimon/popcorn-desktop/tree/readd-peers This is kinda ugly but it works quite well.

Edit: Now that I finished writing this i realize that we probably don't want to destroy the torrent because we want to keep it in the seedbox list in paused state. I added another commit for re-adding peers when resuming from seedbox now.

**Comment by kiriles90** [comment]
From previously testing some of these

1. If that means just relying on `pause()` then.. it takes a very long time (if ever) to stop. When I tried with some good health torrents it kept going full speed for 10-15 minutes before I got bored and aborted it.
2. You can already do that, just remove the torrent without deleting the cache (and you are correct on your edit, we shouldn't rely on `destroy()` for pausing when the Seedbox feature is enabled)
3. That's what I ended up with too as a possible solution but never got around to coding something. Will test https://github.com/tkSimon/popcorn-desktop/tree/readd-peers as soon as possible but from a first look looks good.

**Comment by tkSimon** [comment]
Both changes pushed, but had no time to test it yet. Will probably check tomorrow. For now I created a draft PR

**Comment by tkSimon** [comment]
i updated the PR
